File: java/app/src/main/java/org/example/tds/td3/promowarn/version3/StudentImpl.java
package org.example.td3.promowarn.version3;

import java.util.Objects;
import java.util.Optional;

import org.example.td3.promowarn.common.mail.EMailAddress;

public class StudentImpl implements Student {
    private final Integer id;
    private final EMailAddress email;
    private Double grade;

    public StudentImpl(final Integer id, final String email) {
        this.id = id;
        this.email = new EMailAddress(email);
        this.grade = null; // null au départ !
    }

    @Override
    public Integer id() {
        return id;
    }

    @Override
    public EMailAddress email() {
        return email;
    }

    @Override
    public Optional<Double> grade() {
        return Optional.ofNullable(grade);
    }

    @Override
    public void grade(final double grade) {
        this.grade = grade;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final StudentImpl student = (StudentImpl) o;
        return id == student.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: scala/src/main/scala/tds/td3/impl/StudentImpl.scala
package tds.td3.impl

import tds.td3.common.mail.EMailAddress
import tds.td3.api.Student

// NON FUNCTIONAL: MUTATION
class StudentImpl(override val id: Int, _email: String) extends Student:

    private var _grade: Option[Double] = None

    override val email = EMailAddress(_email)

    override def grade: Option[Double] = _grade

    // NON FUNCTIONAL: MUTATION
    override def setGrade(g: Double): Unit =
        _grade = Option.apply(g)

    override def equals(x: Any): Boolean = 
        if (x == null || getClass!= x.getClass) false
        else id == x.asInstanceOf[StudentImpl].id

    override def hashCode(): Int = id

    override def toString(): String =
        val gradeRepr = grade.map(_.toString).getOrElse("no grade")
        s"$id (email: $email): $gradeRepr"
